runtime/tr.source/trj9/z/codegen/ReduceSynchronizedFieldLoad.cpp

```diff
@@ -66,71 +70,112 @@ ReduceSynchronizedFieldLoad::inlineSynchronizedFieldLoad(TR::Node* node, TR::Cod
-
-      generateSSFInstruction(cg, TR::InstOpCode::LPD, node, registerPair, reusedLoadMemoryReference, lockMemoryReference);
+      generateSSFInstruction(cg, TR::InstOpCode::LPD, node, registerPair, loadMemoryReference, lockMemoryReference);
 
-      if (node->getChild(1)->getOpCodeValue() == TR::l2a)
+      if (loadNode->getOpCodeValue() == TR::l2a)
          {
          generateRREInstruction(cg, TR::InstOpCode::LLGFR, node, loadRegister, loadRegister);
          }
       }
    else
       {
-      TR_ASSERT_SAFE_FATAL((lockWordOffset & 7) == 0 || (generateCompressedLockWord && (lockWordOffset & 3) == 0),
-            "Lockword must be aligned on a double-word boundary or on a word boundary if the lockword is compressed\n");
+      // LPDG requires memory references to be aligned to a double-wrod boundary
+      TR::MemoryReference* alignedLockMemoryReference = lockMemoryReference;
+      TR::MemoryReference* alignedLoadMemoryReference = loadMemoryReference;
 
-      if ((loadMemoryReference->getOffset() & 7) != 0)
-         {
-         TR_ASSERT_SAFE_FATAL((loadMemoryReference->getOffset() & 3) == 0, "Integer field must be aligned on a word boundary\n");
-
-         TR::MemoryReference* alignedLoadMemoryReference = generateS390MemoryReference(*loadMemoryReference, -4, cg);
+      bool lockRegisterRequiresShift = false;
+      bool loadRegisterRequiresShift = false;
 
-         generateSSFInstruction(cg, TR::InstOpCode::LPDG, node, registerPair, alignedLoadMemoryReference, lockMemoryReference);
+      bool lockRegisterRequires32BitLogicalSignExtension = false;
+      bool loadRegisterRequires32BitLogicalSignExtension = false;
 
-         // For 64-bit uncompressed lockwords, when the load memory reference is not aligned to double-word boundaries,
-         // the loaded value is at the lower 32-bit. We need an AND to zero out the high half of loadRegister.
-         if (node->getChild(1)->getOpCodeValue() == TR::iloadi)
+      if (is32BitLock)
+         {
+         if ((lockWordOffset & 7) == 0)
             {
-            generateRILInstruction(cg, TR::InstOpCode::NIHF, node, loadRegister, static_cast<uintptrj_t>(0));
+            lockRegisterRequiresShift = true;
+            }
+         else
+            {
+            // This is because we must use LPDG to load a 32-bit value using displacement -4
+            TR_ASSERT_SAFE_FATAL((lockWordOffset & 3) == 0, "Lockword must be aligned on a word boundary\n");
+
+            lockRegisterRequires32BitLogicalSignExtension = true;
+            alignedLockMemoryReference = generateS390MemoryReference(*lockMemoryReference, -4, cg);
             }
          }
       else
          {
-         TR::MemoryReference* reusedLoadMemoryReference = generateS390MemoryReference(*loadMemoryReference, 0, cg);
-
-         generateSSFInstruction(cg, TR::InstOpCode::LPDG, node, registerPair, reusedLoadMemoryReference, lockMemoryReference);
+         // This is because we must use LPDG to load a 64-bit value
```

Comment:
Not sure why this comment is relevant.. As the assert text says, Lockword must be aligned to DW boundaries anyway.

Reply:
Removed comment in 98f8017